torch/_inductor/codegen/triton.py

```diff
@@ -470,39 +474,39 @@ def _print_Abs(self, expr):
 
     def _print_OpaqueUnaryFn_cos(self, expr):
         assert len(expr.args) == 1
-        return f"libdevice.cos(({self._print(expr.args[0])}).to(tl.float32))"
+        return f"libdevice.cos({self._print(expr.args[0])}{_maybe_upcast_to_fp32()})"
 
     def _print_OpaqueUnaryFn_cosh(self, expr):
         assert len(expr.args) == 1
-        return f"libdevice.cosh(({self._print(expr.args[0])}).to(tl.float32))"
+        return f"libdevice.cosh({self._print(expr.args[0])}{_maybe_upcast_to_fp32()})"
```

Comment:
Do we know if all these `libdevice` calls support fp16? I recall @eellison  had some concerns about that.

Reply:
They dont. we're landing this to unblock mtia and doing more complete dtype analysis later

Reply:
although, cc @arui-meta, maybe we should we could just upcast input to fp32 and then downcast after and that would be sufficient ?

Reply:
As far as I can tell none of these libdevices support lower precision: https://github.com/triton-lang/triton/blob/14025786d108596cfd99700caa4f438938c2ceba/python/triton/language/extra/cuda/libdevice.py#L1192

Reply:
I guess it makes sense to land this as an experimental feature to unblock MTIA. But I'm wondering if these casts for `libdevice` make sense on either arch. MTIA has its own `libdevice` which is different from GPU, but I don't know that it supports all these either.

Reply:
By default we still will upcast to FP32 for GPU, so GPU codegen behavior should stay the same. I'm afraid that we don't have any dtype information on the input args. We don't know what dtype we should downcast to.

Reply:
@arui-meta, you dont have to use compile time dtype, you can use runtime dtype `{expr.args[0]}.dtype`

Reply:
I have added upcast input to fp32 and then downcast to inputs dtype for all the _print_ libdevice functions in the latest version. 

@eellison How about the actual codegen static methods? https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L769 To be more consistent, I think we will need the same upcast on inputs.